### PR TITLE
Fix long string doubled quote parsing

### LINKED ISSUE #71: Inconsistency in double single-quote ('') management in FITS Card
Inconsistency in double single-quote ('') management in FITS Card

### Description

The management of single-quotes in FITS cards seem correct, except sometimes when dealing with null strings, i.e. double single quotes (`''`), which sometimes are transformed into single single quotes (`'`).

E.g.:
```python
from astropy.io import fits
for n in range(60, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''")
    card2 = fits.Card.fromstring(str(card1))  # Should be the same as card1
    print(n, card1.value == card2.value)
    if card1.value != card2.value:
        print(card1.value)
        print(card2.value)
```

Observed output:
```
60 True
61 True
62 True
63 True
64 True
65 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
66 True
67 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
68 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
69 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
```

If the null string `''` is included in a larger value, the issue occurs at a different position:
```python
from astropy.io import fits
for n in range(50, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''" + "x"*10)
    card2 = fits.Card.fromstring(str(card1))
    print(n, len(card1.value), card1.value == card2.value)
```

Observed output:
```
50 62 True
51 63 True
52 64 True
53 65 True
54 66 True
55 67 False
56 68 False
57 69 False
58 70 False
59 71 False
60 72 False
61 73 False
62 74 False
63 75 False
64 76 True
65 77 False
66 78 True
67 79 False
68 80 False
69 81 False
```

Additional cases reported:
- Trailing quote at the end of the line is lost:
```python
fits.Card.fromstring(fits.Card("FOO", "x"*100 + "''", "comment").image).value
# => ends with just one quote
```
- A string after quotes is lost if there is a space in between:
```python
fits.Card.fromstring(fits.Card("FOO", "x"*100 + "'' aaa", "comment").image).value
# => truncates after the single quote
```

### Expected behavior

All card values should be handled properly and round-trip without losing quotes or truncating content.

### How to Reproduce

```python
from astropy.io import fits
for n in range(60, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''")
    card2 = fits.Card.fromstring(str(card1))
    print(n, len(card1.value), card1.value == card2.value)
    if card1.value != card2.value:
        print(card1.value)
        print(card2.value)
```

### Versions

Linux-5.10.0-1029-oem-x86_64-with-glibc2.29
Python 3.8.10 (default, Mar 13 2023, 10:26:41)
[GCC 9.4.0]
astropy 5.2.1
Numpy 1.23.5
pyerfa 2.0.0
Scipy 1.10.0
Matplotlib 3.6.2

---

### Researcher specification

Root cause:
- In astropy/io/fits/card.py, Card._split (long-image path used for continued long string values) performs per-chunk unescaping of doubled single quotes via `replace("''", "'")`. This premature unescape corrupts semantics when doubled quotes are near chunk boundaries or followed by spaces, leading to quote loss or truncation when the reassembled value is parsed.

Proposed fix:
- Do not unescape doubled single quotes per CONTINUE chunk in Card._split. Let the final parse step handle unescaping after full reassembly. Specifically, in Card._split (long-image branch), change:
  - `value = (m.group("strg") or "").rstrip().replace("''", "'")`
  - to: `value = (m.group("strg") or "").rstrip()`
  - Keep existing handling of trailing `&` inside the quoted chunk.
- No changes to serialization logic; it already doubles embedded quotes correctly in `_format_value` and `_format_long_image`.

Optional robustness (defer unless desired):
- Consider tightening the `_strg` regex to `r"'(?P<strg>(?:[^']|'')*)'(?=$|/| )"` to ensure embedded doubled quotes are always treated as content and only a single, non-doubled `'` can terminate the string.

Tests to add (pytest):
- Round-trip for trailing doubled quotes near boundary lengths (e.g., n in 60..100) with and without comments.
- Embedded doubled quotes across boundaries within larger strings; ensure equality after round-trip.
- Reproduce and fix the two additional cases: trailing doubled quote preserved; space-separated text after doubled quotes preserved.
- Null string `''` round-trip, including when in the middle of a long continued value.

Compatibility:
- Change only affects previously buggy cases. Normal behavior is preserved; serialization unchanged.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- retain trailing doubled quotes when splitting CONTINUE segments
- rebuild long string value using full decoded text before re-encoding
- add regression coverage for doubled quote continuations and null strings

## Testing
- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 LD_LIBRARY_PATH=/nix/store/wffgswxkp55xi14jy63rjsnfvl2qvmxy-gcc-14.3.0-lib/lib:  python -m pytest -p pytest_astropy_header -p pytest_doctestplus   /workspace/astropy/.venv/lib/python3.11/site-packages/astropy/io/fits/tests/test_header.py   -k "doubled_quotes or regression_cases_with_trailing_quotes or null_string_round_trip_including_continued_value"

Fixes #71

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Results
- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 LD_LIBRARY_PATH=/nix/store/wffgswxkp55xi14jy63rjsnfvl2qvmxy-gcc-14.3.0-lib/lib:  python -m pytest -p pytest_astropy_header -p pytest_doctestplus   /workspace/astropy/.venv/lib/python3.11/site-packages/astropy/io/fits/tests/test_header.py   -k "doubled_quotes or regression_cases_with_trailing_quotes or null_string_round_trip_including_continued_value" (6 passed, 175 deselected)

**Comment by rowan-stein** [comment]
Requesting review:

This PR fixes long-string FITS Card parsing to preserve doubled single quotes ("''") across CONTINUE segments. Changes:
- Card._split: remove per-chunk unescape of doubled quotes; defer unescape to final parse
- Tests added for boundary lengths, embedded doubled quotes across boundaries, and two reported regression cases

Focus areas:
- Correctness of the long-string reconstruction and unescape flow
- Edge cases with spaces following doubled quotes and trailing quotes at card boundaries

Linked Issue: #71

File: astropy/io/fits/card.py [hunk 1]
```diff
@@ -859,9 +859,29 @@ def _split(self):
...
 
                 value = m.group("strg") or ""
-                value = value.rstrip().replace("''", "'")
+                trailing = vc[m.end("strg") : m.end(0)]
+                if trailing:
+                    trailing_spaces = trailing.replace("'", "")
+                    if trailing_spaces:
+                        space_len = len(trailing_spaces) - len(trailing_spaces.lstrip(" "))
+                        if space_len > 0:
+                            value += trailing_spaces[:space_len]
```
Comment (noa-lucent) [inline astropy/io/fits/card.py:868]: [major] Appending  here breaks the CONTINUE cards that carry the comment. For an input like ,  becomes , so we append the formatting space before the slash to . At that point  is  and the continuation-marker removal below () no longer fires, so we leak a literal  (and the padding space) into . For  we likewise append a stray space chunk. When the fragments are joined we wind up with a user-visible value that ends in  or , so long strings with comments no longer round-trip. Please only recover whitespace that truly belongs to the string payload (e.g. perform this after stripping any trailing  and skip it entirely when a comment is present).

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Summary:\n- The regression tests look comprehensive and they hit the reported reproduction cases.\n- However, the new whitespace-recovery block in  runs before we strip the continuation marker and it is triggered on comment CONTINUE cards. Because of that, we append the formatting space before the slash () into , prevent the  from being stripped, and leak an  fragment into the reassembled string. Any long string with a comment now round-trips with spurious characters.\n\nPlease adjust the trailing-space handling so we only recover genuine value data (e.g. move it after the  trim and skip it when a comment is present) and re-run the affected tests.

**Comment by casey-brooks** [comment]
@noa-lucent Thanks for flagging this. I've reworked the long-card splitter so the ampersand is stripped before any recovery, formatting whitespace is skipped whenever a comment fragment is present, and only genuine value spillover is appended. The regression focus still passes with:

- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 LD_LIBRARY_PATH=/nix/store/wffgswxkp55xi14jy63rjsnfvl2qvmxy-gcc-14.3.0-lib/lib:  python -m pytest -p pytest_astropy_header -p pytest_doctestplus   /workspace/astropy/.venv/lib/python3.11/site-packages/astropy/io/fits/tests/test_header.py   -k "doubled_quotes or regression_cases_with_trailing_quotes or null_string_round_trip_including_continued_value" (6 passed, 175 deselected)

The updated commit is casey-brooks/astropy@6e69ff7. Because of the repository push rules I had to relaunch the PR from my fork (see #101), but the diff there reflects this fix.

**Comment by rowan-stein** [comment]
Superseded by #101

Continuing the work in PR #101 which contains the requested updates. Closing this PR to keep a single active PR for the issue.

Linked Issue: #71
